### PR TITLE
Feature/decrypt

### LINKED ISSUE #125: ✨ Decrypt: compute decryption shares & generate challenge
### Is there an existing issue for this?

- [ ] I have searched the existing issues

### Suggestion

guardians compute decryption shares, admin accumulate tally decryption shares & admin generate shared challenge

### Possible Implementation

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by AddressXception**
[//]: # (🚨 Please review the CONTRIBUTING.md in this repository. 💔Thank you!)

### Issue
*Link your PR to an issue*

Fixes #125

### Description
*Please describe your pull request.*

Adds support for decrypting with shares using the new 2.0 format.

This PR is still a work in progress.
- [x] The Spoiled ballot decryption is failing
- [x] ~Add decryption of hashed elgamal on contests~ (#194)
- [x] ~Add chaum pedersen proofs to verifiable decryption~ (#193)
- [x] Add IsValid checks throughout (see TODOs)
- [x] Clean up code and add documentation

### Testing
*Describe the best way to test or validate your PR.*

1. `make test-netstandard`

### CODE REVIEW

**Comment by AddressXception** [comment]
There's still some work to do on this but I wanted to get it upo so you guys can start a review process since it's another large one.

**Comment by AddressXception** [comment]
Created issues #193 and #194 to complete the proof and write in decryption

File: include/electionguard/chaum_pedersen.hpp [hunk 1]
```diff
@@ -259,6 +259,99 @@ namespace electionguard
...
+        ///                 (usually the accumulated ciphertext decryption, or partial decryption)</param>
+        /// <param name="q">The extended base hash of the election</param>
+        /// <returns> True if everything is consistent. False otherwise. </returns>
+        /// </Summary>
+        bool isValid(const ElGamalCiphertext &message, const ElementModP &k, const ElementModP &m,
+                     const ElementModQ &q);
+
+      private:
+        class Impl;
+#pragma warning(suppress : 4251)
```
Comment (SteveMaier-IRT) [inline include/electionguard/chaum_pedersen.hpp:352]: The other places where we disable this warning do we re-enable it afterwards?

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSecret.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+[TestFixture]
+public class TestDecryptWithSecret : DisposableBase
+{
+    [Test]
+    public void Test_Decrypt_Spoiled_Ballot_With_Nonce()
+    {
+        // Arrange
+        var seed = Constants.ONE_MOD_Q;
+        var nonce = Constants.ONE_MOD_Q;
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSecret.cs:15]: I know it is a unit test, but these will never dispose, as written.

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSecret.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        var manifest = ManifestGenerator.GetFakeManifest();
+        var data = ElectionGenerator.GenerateFakeElectionData(1, 1, manifest);
+
+        var ballot = BallotGenerator.GetFakeBallot(data.InternalManifest);
+
+        // Act
+        var ciphertext = Encrypt.Ballot(
+            ballot, data.InternalManifest, data.Context, seed, nonce);
+        var ballotNonce = new ElementModQ(ciphertext.Nonce);
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSecret.cs:24]: Consider a `using` statement here. again with the leaking test.

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithShares.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    [Test]
+    public void Test_Decrypt_Ballot_With_All_Guardians_Present()
+    {
+        // Arrange
+        var guardians = Data.KeyCeremony.Guardians
+                .ToList();
+
+        var mediator = new DecryptionMediator(
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithShares.cs:85]: Leaksssss

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithShares.cs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+    [Test]
+    public void Test_Decrypt_Ballot_With_Quorum_Guardians_Present()
+    {
+        // Arrange
+        var guardians = Data.KeyCeremony.Guardians
+                .GetRange(0, QUORUM)
+                .ToList();
+
+        var mediator = new DecryptionMediator(
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithShares.cs:117]: leak

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSharesSimple.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            KeyCeremonyGenerator.GenerateKeyCeremonyData(
+            NUMBER_OF_GUARDIANS,
+            QUORUM, runKeyCeremony: true),
+            ManifestGenerator.GetFakeManifest(),
+            BALLOT_COUNT_CAST,
+            BALLOT_COUNT_SPOILED);
+
+        var guardians = data.KeyCeremony.Guardians.ToList();
+
+        var mediator = new DecryptionMediator(
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSharesSimple.cs:33]: 🚰

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSharesSimple.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            QUORUM, runKeyCeremony: true),
+            ManifestGenerator.GetFakeManifest(),
+            BALLOT_COUNT_CAST,
+            BALLOT_COUNT_SPOILED);
+
+        var guardians = data.KeyCeremony.Guardians
+                .GetRange(0, QUORUM)
+                .ToList();
+
+        var mediator = new DecryptionMediator(
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSharesSimple.cs:66]: 🚰

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSharesSimple.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            NUMBER_OF_GUARDIANS,
+            QUORUM, runKeyCeremony: true),
+            ManifestGenerator.GetFakeManifest(),
+            BALLOT_COUNT_CAST,
+            BALLOT_COUNT_SPOILED);
+
+        var guardians = data.KeyCeremony.Guardians
+                .ToList();
+        var spoiledBallots = data.CiphertextBallots.Where(i => i.IsSpoiled).ToList();
+        var mediator = new DecryptionMediator(
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptWithSharesSimple.cs:99]: 🚰

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptionData.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    public PlaintextTally PlaintextTally { get; set; } = default!;
+    public CiphertextTally CiphertextTally { get; set; } = default!;
+
+    // configure a test case by running a key ceremony and generating ballots
+    public static TestDecryptionData ConfigureTestCase(
+        TestKeyCeremonyData keyCeremony,
+        Manifest manifest, int castBallotCount, int spoiledBallotCount)
+    {
+        var internalManifest = new InternalManifest(manifest);
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptionData.cs:28]: 🚰

File: bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptionData.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        var context = new CiphertextElectionContext(
```
Comment (john-s-morgan) [inline bindings/netstandard/ElectionGuard/ElectionGuard.Decryption.Tests/Decryption/TestDecryptionData.cs:29]: 🚰

**Review by john-s-morgan — CHANGES_REQUESTED** [review]
Please clean up the disposable objects before submitting. Lot of potential 🚰. This is a partial list

**Comment by AddressXception** [comment]
PR's are now stacked. memory cleanup is in https://github.com/microsoft/electionguard-core2/pull/205

**Review by SteveMaier-IRT — APPROVED** [review]
looks good to continue with.

**Pull request merged**